accessible/clipboard.service.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * AccessibleBlockly
+ *
+ * Copyright 2016 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Angular2 Service deals with the clipboard and marked spots.
+ * @author madeeha@google.com (Madeeha Ghori)
+ */
+
+var blocklyApp = blocklyApp || {};
+
+blocklyApp.ClipboardService = ng.core
+  .Class({
+    constructor: function() {
+      this.clipboardBlockXml_ = null;
+      this.clipboardBlockSuperiorConnection_ = null;
+      this.clipboardBlockNextConnection_ = null;
+      this.markedConnection_ = null;
+    },
+    cut: function(block) {
+      this.clipboardBlockXml_ = Blockly.Xml.blockToDom(block);
+      this.clipboardBlockSuperiorConnection_ = block.outputConnection ||
+          block.previousConnection;
+      this.clipboardBlockNextConnection_ = block.nextConnection;
+      block.dispose(true);
+      blocklyApp.debug && console.log('cut');
+    },
+    copy: function(block) {
+      this.clipboardBlockXml_ = Blockly.Xml.blockToDom(block);
+      this.clipboardBlockSuperiorConnection_ = block.outputConnection ||
+          block.previousConnection;
+      this.clipboardBlockNextConnection_ = block.nextConnection;
+      blocklyApp.debug && console.log('copy');
+    },
+    paste: function(connection) {
+      var blockOnProperWorkspace = Blockly.Xml.domToBlock(blocklyApp.workspace,
+          this.clipboardBlockXml_);
+      switch (connection.type) {
+        case Blockly.NEXT_STATEMENT:
+          connection.connect(blockOnProperWorkspace.previousConnection);
+          break;
+        case Blockly.PREVIOUS_STATEMENT:
+          connection.connect(blockOnProperWorkspace.nextConnection);
+          break;
+        default:
+          connection.connect(blockOnProperWorkspace.outputConnection);
+      }
+      blocklyApp.debug && console.log('paste');
+    },
+    pasteToMarkedConnection: function(block) {
+      var xml = Blockly.Xml.blockToDom(block);
+      var blockOnProperWorkspace =
+          Blockly.Xml.domToBlock(blocklyApp.workspace, xml);
+      this.markedConnection_.connect(
```

Comment:
pasteFromClipboard knows exactly what it's pasting (the block that's stored in the clipboard), so it only needs to take in the exact connection where the block will be pasted. pasteToMarkedConnection knows exactly which connection the block will be pasted to (the marked connection stored in the clipboard). It doesn't know what block is being connected, so that must be passed in.